docs/runtime/ir-caching.md

```diff
@@ -0,0 +1,209 @@
+out of the equation in its entirety, by serialising the parser's output.
+
+To that end, we want to serialise the Enso IR to a format that can later be read
+back in, bypassing the parser entirely. Furthermore, we can move the boundary at
+which this serialisation takes place to the end of the compiler pipeline,
+thereby bypassing doing most of the compilation work, and further improving
+startup performance.
+
+<!-- MarkdownTOC levels="2,3" autolink="true" -->
+
+- [Serialising the IR](#serialising-the-ir)
+  - [Breaking Links](#breaking-links)
+- [Storing the IR](#storing-the-ir)
+  - [Metadata Format](#metadata-format)
+  - [Portability Guarantees](#portability-guarantees)
+- [Loading the IR](#loading-the-ir)
+  - [Integrity Checking](#integrity-checking)
+  - [Error Handling](#error-handling)
+- [Testing the Serialisation](#testing-the-serialisation)
+
+<!-- /MarkdownTOC -->
+
+## Serialising the IR
+
+As the serialised IR doesn't need to be read by anything other than Enso, we
+need not use a representation that is portable between platforms. As a result,
+we have picked the `Serializable` infrastructure that is _already present_ on
+the JVM. It has the following benefits:
+
+- It is able to serialise arbitrary object graphs while maintaining object
+  identity and tracking references. This cannot be disabled for `Serializable`,
+  but that is fine as we want it.
+- It is built into the JVM and is hence guaranteed to be portable between
+  instances of the same JVM.
+- It copes fine with highly-nested scala types, like our IR.
+
+In order to maximise the benefits of this process, we want to serialise the IR
+as _late_ in the compiler pipeline as possible. This means serialising it just
+before the code generation step that generates Truffle nodes (before the
+`RuntimeStubsGenerator` and `IrToTruffle` run).
+
+This serialisation should take place in an _offloaded thread_ so that
+serialisation doesn't block codegen from occurring.
+
+### Breaking Links
+
+Doing this naïvely, however, means that we can inadvertently end up serialising
+the entire module graph. This is due to the `BindingsMap`, which contains a
+reference to the associated `runtime.Module`, from which there is a reference to
+the `ModuleScope`. The `ModuleScope` may then reference other `runtime.Module`s
+which all contain `IR.Module`s. Therefore, done in a silly fashion, we end up
+serialising the entire reachable module graph. This is not what we want.
+
+While the ideal way of solving this problem would be to customise the
+serialisation and deserialisation process for the `BindingsMap`, the JVM's
+`Serializable` does not provide the ability to customise it enough to solve this
+problem. Instead, we solve it using a preprocessing step:
+
+- We can modify `BindingsMap` and its child types to be able to contain an
+  unlinked module pointer
+  `case class ModulePointer(namespace: String, qualifiedName: List[String])` in
```

Comment:
Why is `namespace` special here?

If `namespace` is special, I would think that the package `name` should also be treated specially, so it would be
`ModulePointer(namespace: String, name: String, qualifiedName: List[String])` where `qualifiedName` would essentially be equivalent to the path of the particular file in the package's `src` directory.